<html lang="en">
    <head>
      <title>Deniz's personal pages</title>
      <meta charset="utf-8">
      <meta name="viewport" content="initial-scale=1">
      <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.13.18/dist/katex.min.css" integrity="sha384-zTROYFVGOfTw7JV7KUu8udsvW2fx4lWOsCEDqhBreBwlHI4ioVRtmIvEThzJHGET" crossorigin="anonymous">
      <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.18/dist/katex.min.js" integrity="sha384-GxNFqL3r9uRJQhR+47eDxuPoNE7yLftQM8LcxzgS4HT73tp970WS/wV5p8UzCOmb" crossorigin="anonymous"></script>
      <script defer src="https://cdn.jsdelivr.net/npm/katex@0.13.18/dist/contrib/auto-render.min.js" integrity="sha384-vZTG03m+2yp6N6BNi5iM4rW4oIwk5DfcNdFfxkk9ZWpDriOkXX8voJBFrAO7MpVl" crossorigin="anonymous"
           onload="renderMathInElement(document.body);"></script>
    </head>
    <body>
        <div>
        <strong>
        <a href="/micro/micro-index.html" onClick="this.href+='?rnd='+new Date().getTime()">Micro</a> | 
        <a href="/post/index.html" onClick="this.href+='?rnd='+new Date().getTime()">Macro</a> | 
        <a href="/post/about.html">About</a> |
        <a href="/post/now.html">Now</a> |
        <a href="/twtxt.txt">twtxt</a>
        </strong>
        </div>
        <hr />
        <div id="main">
        <h1 id="title"> Grails exception mapping with http status codes</h1>
        <div id="date"><em> 2014-10-22</em></div>
        <div class="content">
        <p>an integral part of an API is returning meaningful response codes on errors. If the client is trying to access a restricted resource without authentication the correct code to return would be http 403 etc. Doing this in grails without going through the hassle of try/catching every exception by hand and sending the appropriate code is by using exception mapping feature in grails.</p>
<p>In the <code>UrlMappings</code> class just register the following</p>
<pre><code class="language-groovy">static mappings = {
&quot;/$controller/$action?/$id?(.$format)?&quot;{
   constraints {
       // apply constraints here
   }
}

&quot;/&quot;(view:&quot;/index&quot;)
&quot;500&quot;(controller:'error'),
&quot;500&quot;(controller: 'error', action: 'error403', exception: AuthenticationException)

}
</code></pre>
<p>this tells grails to forward any http 500 error, which are exceptions that you can throw in your code, to the controller and action you define.
For this example we need to define an <code>ErrorController</code> with the action index and <code>error403</code></p>
<pre><code class="language-groovy">class ErrorController {
   def index() {
       response.status = 500
       render([status: &quot;error&quot;, message: &quot;an internal server error occured&quot;])
   }

   def error403() {
       try {
       def message = request.exception?.cause?.message ?: &quot;authentication error&quot;
       response.status = 403
       render([&quot;status&quot;: &quot;error&quot;,&quot;message&quot;: message] as JSON)
       } catch(ex) {
       }
   }
}
</code></pre>
<p>the thing to note here is that</p>
<p>the exceptions that you throw in your code should extend <code>RuntimeException</code>
Even though we catch the exception as a 500 response code in the mappings file, we can override the http response code in the error controller.
you can get the root cause from the request.exception object, which is handy.
it's a good idea to surround the exception handling code in the actions, otherwise if you get an exception in the exception handling code you will find yourself in an infinite loop, and eventually a stack overflow.</p>

        </div>
        </div>
    </body>
</html>
